src/classes/VOL_JRS.cls

```diff
@@ -35,26 +35,64 @@
-	// for the specified JRS's, delete all shifts that have no committed & completed hours.
-	// called from the Before Delete JRS trigger.
-	public static void DeleteListJRS(list<Job_Recurrence_Schedule__c> listJRS) {
-		
-		// get a set of the JRS ID's for querying
-		set<ID> setJRSId = new set<ID>();
-		for (Job_Recurrence_Schedule__c jrs : listJRS) {
-			setJRSId.add(jrs.Id);
-		}
-		
-		// get all shifts associated with these JRS's that we should delete
-		list<Volunteer_Shift__c> listShift = new list<Volunteer_Shift__c>();
-		listShift = [select Id, Total_Volunteers__c from Volunteer_Shift__c  
-			where ((Total_Volunteers__c = null or Total_Volunteers__c = 0) and 
-				Job_Recurrence_Schedule__c in : setJRSId)];
-		
-		delete listShift;
-	}
+    // for the specified JRS's, delete all shifts that have no hours, and detach Shifts that have hours.
+    // called from the Before Delete JRS trigger.
+    public static void DeleteListJRS(list<Job_Recurrence_Schedule__c> listJRS) {
+
+        // get a set of the JRS ID's for querying
+        set<ID> setJRSId = new set<ID>();
+        for (Job_Recurrence_Schedule__c jrs : listJRS) {
+            setJRSId.add(jrs.Id);
+        }
+
+        list<Volunteer_Shift__c> listShiftUpdate = new list<Volunteer_Shift__c>();
+        list<Volunteer_Shift__c> listShiftDelete = new list<Volunteer_Shift__c>();
+        
+        // get all shifts associated with these JRS's
+        for (list<Volunteer_Shift__c> listShift : [select Id, Job_Recurrence_Schedule__r.Name, System_Note__c, 
+            (select Id from Volunteer_Hours__r) 
+            from Volunteer_Shift__c  
+            where Job_Recurrence_Schedule__c in : setJRSId]) {
+            
+            calcDeleteOrDetachJRSShifts(listShift, listShiftDelete, listShiftUpdate);
+        }
+        
+        if (listShiftUpdate.size() > 0)
+            update listShiftUpdate;
+        if (listShiftDelete.size() > 0)
+            delete listShiftDelete;
+    }
 
     //******************************************************************************************************
+    // Given a list of Shifts, add those who have no Hours to the supplied delete list, and those that have
+    // Hours are detached from their JRS (updating their System Note), and put on the update list.
+    // The caller is responsible for committing to the database.
+    private static void calcDeleteOrDetachJRSShifts(
+        list<Volunteer_Shift__c> listShift, 
+        list<Volunteer_Shift__c> listShiftDelete, 
+        list<Volunteer_Shift__c> listShiftUpdate) {
+
+        for (Volunteer_Shift__c shift : listShift) {
+            Integer count = 0;
+            // note that to avoid query limits, we avoid looking at Volunteer_Hours__r.size(), and 
+            // must leverage soql for loops that internally use queryMore. 
+            for (Volunteer_Hours__c hr : shift.Volunteer_Hours__r) {
```

Comment:
I was hoping there would be some standard rollups of hours to shifts, so we could just use that field and avoid the subqueries altogether, but alas, there aren't any.